### PR TITLE
Make node-sass typings for the public API complete and more accurate.

### PR DESCRIPTION
**Pull request by chriseppstein**
## Description

This patch adds types for the [javascript representation of sass functions][function-api] implemented in JS and the data structures that are passed to and returned from those functions.
types. Building on that, there are updates to the definitions of the importer and functions options to be more specific and avoid the use of `any` for arguments passed to functions from node-sass.

The [context objects][ctx] and async APIs are bifurcated so that it is now illegal to pass async functions to a synchronous compilation.

This patch is a likely breaking change for anyone who's using types with node-sass, especially if they're working with importers or function declarations. Because node-sass is now at 4.11, I've recommend a major and minor version bump to match the latest version of node-sass
`4.11.0`. This will prevent existing users of @types/node-sass from breaking.

### Checklist

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.) https://github.com/sass-eyeglass/eyeglass/pull/202
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [API Docs for Sass Extension Functions][function-api] _Note: most of this API was already in existence when these types were published, this patch just fleshes it out._
- [x] Increase the version number in the header if appropriate. ⬆️ `4.11.0`
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. _already done_
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`. _I enabled `strictNullChecks`_

### Potential Reviewers:
* @giladgray did some work on this in the past.
* I think @pspeter3 did the initial typings for node-sass for his work at asana.
* @xzyfer is the node-sass lead dev. Perhaps he can help review the public API and confirm any questions about it.

[function-api]: https://github.com/sass/node-sass#functions--v300---experimental
[ctx]: https://github.com/sass/node-sass/blob/master/lib/index.js#L209

### CODE REVIEW

File: types/node-sass/index.d.ts [hunk 1]
```diff
@@ -1,56 +1,398 @@
...
+
+export interface AsyncContext extends Context {
+	callback: SassRenderCallback;
+}
+
+export interface SyncContext extends Context {
+	callback: undefined;
+}
+
+export type AsyncImporter = (this: AsyncContext, url: string, prev: string, done: (data: ImporterReturnType) => void) => void;
```
Comment (pspeter3) [inline types/node-sass/index.d.ts]: This looks a lot better.

**Review by pspeter3 — COMMENTED** [review]
This looks good as far as I can tell. Thank you for updating the typings!

File: types/node-sass/index.d.ts [hunk 2]
```diff
@@ -1,56 +1,398 @@
-// Type definitions for Node Sass v3.10.1
+// Type definitions for node-sass 4.11.0
```
Comment (DanielRosenwasser) [inline types/node-sass/index.d.ts]: Looks like our lint rules are disabled for this package. Can you enable the lints in our `tslint.json` and fix up the version string to only be a `MAJOR.MINOR` version (e.g. `4.7`), as opposed to including the patch or containing a leading `v` (e.g. `v1.2.3`).

Comment (DanielRosenwasser) [reply types/node-sass/index.d.ts]: If re-enabling too many rules is cumbersome, as many as is reasonable would go a long way.

Comment (chriseppstein) [reply types/node-sass/index.d.ts]: @DanielRosenwasser Thanks for reviewing! I've sorted out the build and re-enabled all the linters except for one. https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33017/commits/ad1f840706fde09ed2da1ef83f19111b22381223

**Comment by DanielRosenwasser** [comment]
Thanks Chris!

**Pull request merged**